src/feat/online-feature.cc

```diff
@@ -35,26 +35,63 @@ template<class C>
 OnlineGenericBaseFeature<C>::OnlineGenericBaseFeature(
     const typename C::Options &opts):
     computer_(opts), window_function_(computer_.GetFrameOptions()),
-    input_finished_(false), waveform_offset_(0) { }
+    input_finished_(false), waveform_offset_(0),
+    waveform_downsampler_(NULL) { }
 
 template<class C>
 void OnlineGenericBaseFeature<C>::AcceptWaveform(BaseFloat sampling_rate,
                                                  const VectorBase<BaseFloat> &waveform) {
   BaseFloat expected_sampling_rate = computer_.GetFrameOptions().samp_freq;
-  if (sampling_rate != expected_sampling_rate)
-    KALDI_ERR << "Sampling frequency mismatch, expected "
-              << expected_sampling_rate << ", got " << sampling_rate;
-  if (waveform.Dim() == 0)
+  bool allow_downsample = computer_.GetFrameOptions().allow_downsample;
+  Vector<BaseFloat> downsampled_wave;
+  int32 waveform_dim = waveform.Dim();
+  if (sampling_rate < expected_sampling_rate)
+    KALDI_ERR << "The waveform is allowed to get downsampled."
+              << "New sample Frequency " << expected_sampling_rate
+              << " is larger than waveform original sampling frequency "
+              << sampling_rate;
+  else if (expected_sampling_rate < sampling_rate) {
+    if (! allow_downsample)
+      KALDI_ERR << "Waveform and config sample Frequency mismatch: "
+                << sampling_rate << " .vs " << expected_sampling_rate
+                << " ( use --allow_downsample=true option to allow "
+                << " downsampling the waveform).";
+
+    // Downsample the waveform
+    if (waveform_downsampler_ == NULL) {
+      BaseFloat lowpass_cutoff = 0.99 * 0.5 * expected_sampling_rate;
+      int32 lowpass_filter_width = 6;
+      waveform_downsampler_ = new LinearResample(sampling_rate,
+                                                 expected_sampling_rate,
+                                                 lowpass_cutoff,
+                                                 lowpass_filter_width);
+    }
+    BaseFloat old_sampling_rate = waveform_downsampler_->InputSampleRate();
+    // Check wheter the sampling rate values provided on different calls are
+    // the same.
+    if (sampling_rate != old_sampling_rate)
+      KALDI_ERR << "Waveform sampling rate value provided in this call "
+                << "mismatches the value in older calls:"
+                << sampling_rate << " .vs " << old_sampling_rate;
+
+    // flush out the last few samples of input waveform, if input_finished_
+    // is true
+    const bool flush = input_finished_;
```

Comment:
AcceptWaveform() should never be called if input_finished_ is false.  It will always be false.